cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java

```diff
@@ -0,0 +1,195 @@
+     * handler that sets up the principal for the CWMS Access Manager will likely
+     * have easier workflow.
+     * @param ctx
+     * @param user
+     * @param key the key
+     */
+    private void prepareContextWithUser(Context ctx, String user,String key) throws SQLException {
+        logger.info("Validated Api Key for user=" + user);
+
+        ConnectionPreparer keyPreparer = new ApiKeyUserPreparer(key);
+        ConnectionPreparer officePrepare = new SessionOfficePreparer(ctx.queryParam("office"));
+        //ConnectionPreparer resetPreparer = new DirectUserPreparer("q0webtest");
+        DelegatingConnectionPreparer apiPreparer = new DelegatingConnectionPreparer(officePrepare,keyPreparer);
+
+        if (dataSource instanceof ConnectionPreparingDataSource) {
+            ConnectionPreparingDataSource cpDs = (ConnectionPreparingDataSource)    dataSource;
+            ConnectionPreparer existingPreparer = cpDs.getPreparer();
+
+            // Have it do our extra step last.
+            cpDs.setPreparer(new DelegatingConnectionPreparer(existingPreparer, apiPreparer));
+        } else {            
+            ctx.attribute(ApiServlet.DATA_SOURCE, new ConnectionPreparingDataSource(apiPreparer, dataSource));
+        }
+    }
+
+    private String authorized(Context ctx, String key, Set<RouteRole> routeRoles)
+    {
+        String retval = null;
+        String user = checkKey(key,ctx);
+
+        if (routeRoles == null || routeRoles.isEmpty()) {
+            retval = user;
+        } else {
+            Set<RouteRole> specifiedRoles = getRoles(user,ctx.queryParam("office"));
+            retval = specifiedRoles.containsAll(routeRoles) == true ? user : null;
+        }
+
+        return retval;
+    }
+
+    private Set<RouteRole> getRoles(String user,String office) {
+        Set<RouteRole> roles = new HashSet<>();
+        try(Connection conn = dataSource.getConnection();
+            PreparedStatement getRoles = conn.prepareStatement(RETRIEVE_GROUPS_OF_USER);
+        ) {
+            getRoles.setString(1,user);
+            getRoles.setString(2,office);
+            try (ResultSet rs = getRoles.executeQuery()) {
+                while(rs.next()) {
+                    roles.add(new Role(rs.getString(1)));
+                }
+            }
+        } catch(SQLException ex) {
+            logger.log(Level.WARNING,"Failed to retrieve roles for user",ex);
+        }
+        return roles;
+    }
+
+    private String checkKey(String key, Context ctx) {
+        try(Connection conn = dataSource.getConnection();
+            PreparedStatement setApiUser = conn.prepareStatement("begin cwms_env.set_session_user_direct(upper('q0hectest_pu')); end;");
+            PreparedStatement checkForKey = conn.prepareStatement("select userid from cwms_20.at_api_keys where apikey = ?")) 
+        {
+            setApiUser.execute();
+            checkForKey.setString(1,key);
```

Comment:
I'll be curious to see all the db interactions that are happening for each request

Reply:
So far it's seemed pretty lightweight, if it starts to add up we can implement a form of session system were anyone can use what ever credentials have creds that don't need to round trip to the database for basic checks.